source/credentials_provider_sso.c

```diff
@@ -0,0 +1,843 @@
+            .expiration_required = true,
+            .expiration_format = AWS_PCEF_NUMBER_UNIX_EPOCH_MS,
+        };
+
+        credentials = aws_parse_credentials_from_json_document(
+            sso_query_context->allocator, aws_byte_cursor_from_buf(&sso_query_context->payload), &parse_options);
+    }
+
+    if (credentials) {
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) successfully queried credentials",
+            (void *)sso_query_context->provider);
+    } else {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) failed to query credentials",
+            (void *)sso_query_context->provider);
+
+        if (sso_query_context->error_code == AWS_ERROR_SUCCESS) {
+            sso_query_context->error_code = AWS_AUTH_CREDENTIALS_PROVIDER_SSO_SOURCE_FAILURE;
+        }
+    }
+
+    /* pass the credentials back */
+    sso_query_context->original_callback(
+        credentials, sso_query_context->error_code, sso_query_context->original_user_data);
+
+    /* clean up */
+    s_sso_query_context_destroy(sso_query_context);
+    aws_credentials_release(credentials);
+}
+static void s_on_retry_ready(struct aws_retry_token *token, int error_code, void *user_data);
+
+static void s_on_stream_complete_fn(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct aws_sso_query_context *sso_query_context = user_data;
+
+    struct aws_credentials_provider_sso_impl *impl = sso_query_context->provider->impl;
+    impl->function_table->aws_http_stream_release(stream);
+
+    /* set error code */
+    sso_query_context->error_code = error_code;
+    impl->function_table->aws_http_stream_get_incoming_response_status(stream, &sso_query_context->status_code);
+    if (error_code == AWS_OP_SUCCESS && sso_query_context->status_code != AWS_HTTP_STATUS_CODE_200_OK) {
+        sso_query_context->error_code = AWS_AUTH_CREDENTIALS_PROVIDER_HTTP_STATUS_FAILURE;
+    }
+
+    /*
+     * If we can retry the request based on error response or http status code failure, retry it, otherwise, call the
+     * finalize function.
+     */
+    if (error_code || sso_query_context->status_code != AWS_HTTP_STATUS_CODE_200_OK) {
+        enum aws_retry_error_type error_type =
+            aws_credentials_provider_compute_retry_error_type(sso_query_context->status_code, error_code);
+
+        /* don't retry client errors at all. */
+        if (error_type != AWS_RETRY_ERROR_TYPE_CLIENT_ERROR) {
+            if (aws_retry_strategy_schedule_retry(
+                    sso_query_context->retry_token, error_type, s_on_retry_ready, sso_query_context) ==
+                AWS_OP_SUCCESS) {
+                AWS_LOGF_INFO(
+                    AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                    "(id=%p): successfully scheduled a retry",
+                    (void *)sso_query_context->provider);
+                return;
+            }
+            AWS_LOGF_ERROR(
```

Comment:
for this error path, should we set `so_query_context->error_code = aws_last_error();`?

Reply:
thanks, fixed.